packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -493,6 +496,24 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
           },
       );
 
+      registerServiceExtension(name: 'invokePreHotRestartCallbacks', callback: (Map<String, Object> params) async {
+        Future<void> invokeAndWait(PreHotRestartCallback callback, String? label) async {
+          try {
+            await callback();
+          } catch (error) {
+            debugPrint('Failed to invoke preHotRestartCallback ${label ?? 'unlabeled'}: $error');
+          }
+        }
+
+        await Future.wait(<Future<void>>[
+          for (MapEntry<PreHotRestartCallback, String?> entry in _hotRestartCallbacks.entries)
+            invokeAndWait(entry.key, entry.value).timeout(const Duration(seconds: 5), onTimeout: () {
```

Comment:
If we have timeouts here, we really should do them on the tool side... but I would just remove the timeouts entirely.

Optionally, we could make a second hot restart request while one is already ongoing just force a hot restart without triggering these callbacks.

Reply:
I can remove the timeouts here, possibly use the VM service notification API to let the tool know which callbacks are pending. then the timeout warning could include what exactly is stuck.

Rather than a second hot restart, we could prompt to continue ignoring the callbacks as long as a terminal is attached. If one of these is caught in an infinite loop though, there isn't anything we can do (though that is true of other parts of the framework too: https://github.com/flutter/flutter/issues/85987)

Reply:
Prompting requires more work to make this work in IDEs. Forcing the hot restart is always what you're going to want, as a developer. When would you ever say "no, I'd rather not do the thing I asked for actually, just continue waiting on this thing that's never going to finish"? :-)